virtualBridge.py

```diff
@@ -34,40 +36,63 @@ def __init__(self, h: host.Host, config: BridgeConfig):
-        # with wrong mac -> remove it
-
-        cmd = "virsh net-dumpxml default"
-        ret = self.hostconn.run_or_die(cmd)
-        if f"'{name}'" in ret.out:
-            logger.info(f"{name} already configured as static DHCP entry - removing before adding back with proper configuration")
-            host_xml = f"<host name='{name}'/>"
-            cmd = f"virsh net-update default delete ip-dhcp-host \"{host_xml}\" --live --config"
+    def setup_dhcp_entries(self, vms: list[NodeConfig]) -> None:
+        # DHCP entries should have been removed during teardown.
+        # However, leases sometimes came back.
+        self.remove_dhcp_entries(vms)
+        for cfg in vms:
+            if cfg.ip is None:
+                logger.error_and_exit(f"Missing IP for node {cfg.name}")
+            ip = cfg.ip
+            mac = cfg.mac
+            name = cfg.name
+
+            host_xml = f"<host mac='{mac}' name='{name}' ip='{ip}'/>"
+            logger.info(f"Creating static DHCP entry for VM {name}, ip {ip} mac {mac}")
+            cmd = f"virsh net-update default add ip-dhcp-host \"{host_xml}\" --live --config"
             self.hostconn.run_or_die(cmd)
 
-        cmd = "virsh net-dhcp-leases default"
-        ret = self.hostconn.run(cmd)
-        # Look for "{name} " in the output. The space is intended to differentiate between "bm-worker-2 " and e.g. "bm-worker-20"
-        if f"{name} " in ret.out:
-            logger.error(f"Error: {name} found in dhcp leases")
-            logger.error("To fix this, run")
-            logger.error("\tvirsh net-destroy default")
-            logger.error("\tRemove wrong entries from /var/lib/libvirt/dnsmasq/virbr0.status")
-            logger.error("\tvirsh net-start default")
-            logger.error("\tsystemctl restart libvirt")
-            sys.exit(-1)
-
-        host_xml = f"<host mac='{mac}' name='{name}' ip='{ip}'/>"
-        logger.info(f"Creating static DHCP entry for VM {name}, ip {ip} mac {mac}")
-        cmd = f"virsh net-update default add ip-dhcp-host \"{host_xml}\" --live --config"
-        self.hostconn.run_or_die(cmd)
+    def remove_dhcp_entries(self, vms: list[NodeConfig]) -> None:
+        xml_str = self.hostconn.run("virsh net-dumpxml default").out
+        q = et.fromstring(xml_str)
+        removed_macs = []  # type: list[str]
+        names = [x.name for x in vms]
+        ips = [x.ip for x in vms]
+        for e in q[-1][0][1:]:
+            if e.attrib["name"] in names or e.attrib["ip"] in ips:
+                mac = e.attrib["mac"]
+                name = e.attrib["name"]
+                ip = e.attrib["ip"]
+                pre = "virsh net-update default delete ip-dhcp-host"
+                cmd = f"{pre} \"<host mac='{mac}' name='{name}' ip='{ip}'/>\" --live --config"
+                logger.info(self.hostconn.run(cmd))
+                removed_macs.append(mac)
+
+        fn = "/var/lib/libvirt/dnsmasq/virbr0.status"
```

Comment:
we want to python path library for this purpose

Reply:
Will do in updated PR.